File: Threema/SharedUI/Utilities/DocumentPicker.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2015-2024 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "DocumentPicker.h"
#import "UTIConverter.h"
#import "ModalPresenter.h"
#import "BundleUtil.h"
#import <MBProgressHUD/MBProgressHUD.h>
#import "ContactUtil.h"
#import "Old_FileMessageSender.h"
#import <Contacts/Contacts.h>
#import <ContactsUI/ContactsUI.h>
#import "UserSettings.h"
#import "Threema-Swift.h"

@import CocoaLumberjack;

@interface DocumentPicker () <UIDocumentPickerDelegate, UIDocumentMenuDelegate, UploadProgressDelegate, CNContactPickerDelegate>

@property UIViewController *presentingViewController;
@property Conversation *conversation;

@end

@implementation DocumentPicker

static DocumentPicker *pickerStrongReference;

+ (instancetype)documentPickerForViewController:(UIViewController *)presentingViewController conversation:(Conversation *)conversation {
    DocumentPicker *picker = [[DocumentPicker alloc] init];
    picker.presentingViewController = presentingViewController;
    picker.conversation = conversation;
    
    return picker;
}

- (void)show {
    [self showPicker];
}

- (void)showPicker {
    NSArray *types = @[UTTYPE_ITEM, UTTYPE_DATA, UTTYPE_CONTENT, UTTYPE_ARCHIVE, UTTYPE_CONTACT, UTTYPE_MESSAGE];
    
    UIDocumentMenuViewController *controller = [[UIDocumentMenuViewController alloc] initWithDocumentTypes:types inMode:UIDocumentPickerModeImport];
    controller.delegate = self;

    NSString *title = [BundleUtil localizedStringForKey:@"contacts"];
    UIImage *image = [BundleUtil imageNamed:@"ThumbBusinessContact.png"];
    [controller addOptionWithTitle:title image:image order:UIDocumentMenuOrderFirst handler:^{
        [self checkPermissionAndShowContactPicker];
    }];
    
    pickerStrongReference = self;
        
    [self showController:controller];
}

- (void)showController:(UIViewController *)controller {
    if (CGRectIsNull(_popoverSourceRect)) {
        [ModalPresenter present:controller on:_presentingViewController];
    } else {
        [ModalPresenter present:controller on:_presentingViewController fromRect:_popoverSourceRect inView:_presentingViewController.view];
    }
}

- (void)checkPermissionAndShowContactPicker {
    if ([CNContactStore authorizationStatusForEntityType:CNEntityTypeContacts] == CNAuthorizationStatusAuthorized) {
        [self showContactPicker];
    }
    else if ([CNContactStore authorizationStatusForEntityType:CNEntityTypeContacts] == CNAuthorizationStatusDenied) {
        [self showContactAlert];
    }
    else {
        CNContactStore *cnAddressBook = [CNContactStore new];
        [cnAddressBook requestAccessForEntityType:CNEntityTypeContacts completionHandler:^(BOOL granted, NSError * _Nullable error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (granted) {
                    [self showContactPicker];
                } else {
                    [self showContactAlert];
                }
            });
        }];
    }
}

- (void)showContactPicker {
    CNContactPickerViewController *picker = [[CNContactPickerViewController alloc] init];
    picker.delegate = self;
    picker.modalPresentationStyle = UIModalPresentationFormSheet;
    [self showController:picker];
}

- (void)showContactAlert {
    // Show access prompt
    [UIAlertTemplate showOpenSettingsAlertWithOwner:_presentingViewController noAccessAlertType:NoAccessAlertTypeContacts];
}

- (void)sendItem:(URLSenderItem *)item {
    NSString *messageFormat = [BundleUtil localizedStringForKey:@"send_file_message"];
    NSString *message = [NSString stringWithFormat:messageFormat, [item getName], _conversation.displayName];
    NSString *title = [BundleUtil localizedStringForKey:@"send_file_title"];
    __block UITextField *captionTextField = nil;
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = [BundleUtil localizedStringForKey:@"optional_caption"];
        captionTextField = textField;
    }];
    NSString *ok = [BundleUtil localizedStringForKey:@"ok"];
    NSString *cancel = [BundleUtil localizedStringForKey:@"cancel"];
    UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:ok style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {
        
        if (captionTextField.text.length > 0) {
            item.caption = captionTextField.text;
        }
        
        if (_conversation != nil) {
            MessageSender *messageSender = [[BusinessInjector new] messageSenderObjC];
            [messageSender sendBlobMessageFor:item inConversationWithID:_conversation.objectID correlationID:nil webRequestID:nil completion:nil];
        } else {
            [NotificationPresenterWrapper.shared presentSendingError];
        }
    }];
    [alertController addAction:defaultAction];
    
    UIAlertAction* cancelAction = [UIAlertAction actionWithTitle:cancel style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
        pickerStrongReference = nil;
    }];
    [alertController addAction:cancelAction];
    
    [_presentingViewController presentViewController:alertController animated:YES completion:nil];
}

#pragma mark - UIDocumentPickerDelegate

- (void)documentPicker:(UIDocumentPickerViewController *)controller didPickDocumentAtURL:(NSURL *)url {
    NSString *mimeType = [UTIConverter mimeTypeFromUTI:[UTIConverter utiForFileURL:url]];
    URLSenderItem *item = [URLSenderItem itemWithUrl:url type:mimeType renderType:@0 sendAsFile:true];
    [self sendItem:item];
}

#pragma mark - UIDocumentMenuDelegate

- (void)documentMenu:(UIDocumentMenuViewController *)documentMenu didPickDocumentPicker:(UIDocumentPickerViewController *)documentPicker {
    documentPicker.delegate = self;
    
    [self showController:documentPicker];
}

#pragma mark - UploadProgressDelegate

- (BOOL)blobMessageSenderUploadShouldCancel:(Old_BlobMessageSender *)blobMessageSender {
    return NO;
}

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadProgress:(NSNumber *)progress forMessage:(BaseMessage *)message {
    // hide as soon as progress starts which is visible in message bubble
    /// Progress might not be reported in note to self groups. To make sure the HUD is actually dimissed it will be dismissed again in `- (void)blobMessageSender:(BlobMessageSender *)blobMessageSender uploadSucceededForMessage:(BaseMessage *)message`
    [MBProgressHUD hideHUDForView:_presentingViewController.view animated:YES];
}

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadFailedForMessage:(BaseMessage *)message error:(UploadError)error {
    [MBProgressHUD hideHUDForView:_presentingViewController.view animated:YES];
    
    NSString *errorTitle = [BundleUtil localizedStringForKey:@"error_sending_failed"];
    NSString *errorMessage = [Old_FileMessageSender messageForError:error];
    [self showAlertWithTitle:errorTitle message:errorMessage closeOnOk:NO];
}

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadSucceededForMessage:(BaseMessage *)message {
    // Upload succeeds immediately when sending files in note groups without ever reporting upload progress
    [MBProgressHUD hideHUDForView:_presentingViewController.view animated:YES];
    pickerStrongReference = nil;
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message closeOnOk:(BOOL)closeOnOk {
    [UIAlertTemplate showAlertWithOwner:_presentingViewController title:title message:message actionOk:^(UIAlertAction * _Nonnull okAction) {
        pickerStrongReference = nil;
    }];
}


#pragma mark - People picker delegate

-(void) contactPicker:(CNContactPickerViewController *)picker didSelectContact:(CNContact *)contact{
    NSData *vCardData = [ContactUtil vCardDataForCnContact:contact];
    URLSenderItem *item = [URLSenderItem itemWithData:vCardData fileName:nil type:UTTYPE_VCARD renderType:@0 sendAsFile:true];
    
    [picker dismissViewControllerAnimated:YES completion:^{
        [self sendItem:item];
    }];
}

- (void)contactPicker:(CNContactPickerViewController *)picker didSelectContactProperty:(CNContactProperty *)contactProperty  {
    [picker dismissViewControllerAnimated:YES completion:nil];
}

-(void)contactPickerDidCancel:(CNContactPickerViewController *)picker {
    [picker dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Threema/Chats/ChatView/DocumentPicker.swift
import ContactsUI
import FileUtility
import MobileCoreServices
import ThreemaMacros
import UIKit
import UniformTypeIdentifiers

final class DocumentPicker: NSObject {
    
    // MARK: - Properties
    
    private lazy var messageSender = BusinessInjector.ui.messageSender
    
    private let conversation: ConversationEntity
    private weak var presenter: UIViewController?

    // MARK: - Lifecycle
    
    init(for conversation: ConversationEntity, presenter: UIViewController?) {
        self.conversation = conversation
        self.presenter = presenter
    }
    
    // MARK: - Show pickers
    
    func checkPermissionAndShowContactPicker() {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        let isAuthorizedOrLimited: Bool =
            if #available(iOS 18.0, *) {
                status == .authorized || status == .limited
            }
            else {
                status == .authorized
            }

        if isAuthorizedOrLimited {
            showContactPicker()
        }
        else if status == .denied {
            showContactAlert()
        }
        else {
            CNContactStore().requestAccess(for: .contacts) { granted, _ in
                DispatchQueue.main.async {
                    granted ? self.showContactPicker() : self.showContactAlert()
                }
            }
        }
    }

    private func showContactPicker() {
        let picker = CNContactPickerViewController()
        picker.delegate = self
        presenter?.present(picker, animated: true)
    }

    func showDocumentPicker() {
        let allowedTypes: [UTType] = [.item, .data, .content, .archive, .contact, .message]
        let picker = UIDocumentPickerViewController(forOpeningContentTypes: allowedTypes)
        picker.delegate = self
        picker.allowsMultipleSelection = false
        presenter?.present(picker, animated: true)
    }
    
    // MARK: - Helpers
    
    private func showContactAlert() {
        guard let presenter else {
            return
        }
        
        UIAlertTemplate.showOpenSettingsAlert(
            owner: presenter,
            noAccessAlertType: .contacts
        )
    }
    
    private func sendItem(_ item: URLSenderItem) {
        let messageFormat = #localize("send_file_message")
        let message = String(format: messageFormat, item.getName(), conversation.displayName)
        let title = #localize("send_file_title")

        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alertController.addTextField { textField in
            textField.placeholder = #localize("optional_caption")
        }

        let sendAction = UIAlertAction(title: #localize("send"), style: .default) { [weak self] _ in
            guard let self else {
                return
            }
            
            let caption = alertController.textFields?.first?.text ?? ""
            if !caption.isEmpty {
                item.caption = caption
            }

            sendBlobMessage(for: item, in: conversation.objectID)
        }

        let cancelAction = UIAlertAction(title: #localize("cancel"), style: .cancel)

        alertController.addAction(sendAction)
        alertController.addAction(cancelAction)

        presenter?.present(alertController, animated: true)
    }
    
    private func sendBlobMessage(for item: URLSenderItem, in conversationObjectID: NSManagedObjectID) {
        guard let presenter else {
            return
        }

        Task {
            do {
                try await messageSender.sendBlobMessage(
                    for: item,
                    in: conversationObjectID,
                    correlationID: nil,
                    webRequestID: nil
                )
            }
            catch let error as MessageSenderError {
                Task { @MainActor in
                    let title = #localize("error_sending_failed")
                    let message: String
                    switch error {
                    case .tooBig:
                        let format = #localize("error_message_file_too_big")
                        message = String(
                            format: format,
                            arguments: [FileUtility.shared.getFileSizeDescription(from: Int64(kMaxFileSize))]
                        )

                    case .noData:
                        message = #localize("error_message_invalid_file")

                    default:
                        message = #localize("error_message_generic")
                    }

                    UIAlertTemplate.showAlert(
                        owner: presenter,
                        title: title,
                        message: message,
                    )
                }
            }
            catch let error as BlobManagerError {
                guard error != .noteGroupNeedsNoSync else {
                    return
                }
                Task { @MainActor in
                    let title = #localize("error_sending_failed")
                    let message = error.localizedDescription

                    UIAlertTemplate.showAlert(
                        owner: presenter,
                        title: title,
                        message: message,
                    )
                }
            }
            catch {
                Task { @MainActor in
                    let title = #localize("error_sending_failed")
                    let message = error.localizedDescription

                    UIAlertTemplate.showAlert(
                        owner: presenter,
                        title: title,
                        message: message,
                    )
                }
            }
        }
    }
}

// MARK: - CNContactPickerDelegate

extension DocumentPicker: CNContactPickerDelegate {
    
    // MARK: - Delegate functions
    
    func contactPicker(_ picker: CNContactPickerViewController, didSelect contact: CNContact) {
        guard let vCardData = ContactUtil.vCardData(for: contact) else {
            return
        }

        guard let item = URLSenderItem(
            data: vCardData,
            fileName: nil,
            type: UTType.vCard.identifier,
            renderType: 0,
            sendAsFile: true
        ) else {
            return
        }

        picker.dismiss(animated: true) { [weak self] in
            self?.sendItem(item)
        }
    }
}

// MARK: - UIDocumentPickerDelegate

extension DocumentPicker: UIDocumentPickerDelegate {
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let url = urls.first else {
            return
        }

        guard url.startAccessingSecurityScopedResource() else {
            return
        }
        
        defer {
            url.stopAccessingSecurityScopedResource()
        }
        
        guard let data = try? Data(contentsOf: url) else {
            return
        }

        let uti = UTIConverter.uti(forFileURL: url) ?? UTType.data.identifier
        guard let item = URLSenderItem(
            data: data,
            fileName: url.lastPathComponent,
            type: uti,
            renderType: 0,
            sendAsFile: true
        ) else {
            return
        }

        sendItem(item)
    }
}
